File: TypeScript/merge_two_sorted_lists/21.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
  const head = new ListNode();
  let curr = head,
    p1 = list1,
    p2 = list2;

  while (p1 !== null && p2 !== null) {
    if (p1.val < p2.val) {
      curr.next = p1;
      p1 = p1.next;
    } else {
      curr.next = p2;
      p2 = p2.next;
    }
    curr = curr.next;
  }
  curr.next = p1 ?? p2;

  return head.next;
}

export { ListNode, mergeTwoLists };

File: Golang/merge_two_sorted_lists/21.go
package mergetwosortedlists

import (
	helpers "leetcode"
)

type ListNode = helpers.ListNode

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	head := ListNode{}
	prev := &head

	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			prev.Next = list1
			list1 = list1.Next
		} else {
			prev.Next = list2
			list2 = list2.Next
		}
		prev = prev.Next
	}

	if list1 != nil {
		prev.Next = list1
	} else {
		prev.Next = list2
	}

	return head.Next
}
